### PR TITLE
PostgreSQL Backend

### PR DESCRIPTION
**Pull request by cpcloud**

### CODE REVIEW

**Comment by cpcloud** [comment]
A lot of these tests are shamelessly copy pasted from the sqlite backend. I'd like to refactor them into a base set of tests.

**Comment by wesm** [comment]
this is awesome! thank you for getting it started. I agree that having a base test suite for all dialects (and a mechanism to "turn off" tests for not-implemented functionality) is the right path to avoid some of the code duplication. I can help with that

I'm also thinking it would be useful to auto-generate a "feature matrix" to show which operations are implemented (either as rewrites or translation rules) by each compiler

**Comment by wesm** [comment]
Will need to add a reproducible test data load. It would not be difficult to modify `scripts/test_data_admin.py` to have a postgres mode and some parameters. I can look into that if you like

**Comment by cpcloud** [comment]
Ping! I've significantly expanded the original PR. We now have most string functions and window functions (working on cumulative today and tomorrow if I don't finish it)

@wesm Can you give the test data gen script a try with the new additions to load data into posgres? I don't have any HDFS machines to test this on.

**Comment by wesm** [comment]
Awesome! Let me try to do this sometime today, will report back

**Comment by wesm** [comment]
I wasn't able to look at this before going on vacation; I will try out the test suite to make sure everything's A-OK and look to merge within the next week. Sorry for the delay, looking forward to pushing forward the pgsql backend

**Comment by cpcloud** [comment]
Cool dude, let me know if anything needs fixin'

**Comment by cpcloud** [comment]
Happy to squash as desired, if necessary.

**Comment by cpcloud** [comment]
Added the final touch: support for cumulative window functions.

File: ibis/sql/alchemy.py [hunk 1]
```diff
@@ -894,3 +911,61 @@ def _floor_divide(t, expr):
...
+    # We need to define this for window functions that have an order by
+    by, ascending = expr.op().args
+    sort_direction = sa.asc if ascending else sa.desc
+    return sort_direction(t.translate(by))
+
+
+_valid_frame_types = numbers.Integral, str, type(None)
+
+
+class Over(_Over):
```
Comment (cpcloud) [inline ibis/sql/alchemy.py]: I have a PR to SQLAlchemy to add frame specification to `Over` elements: https://github.com/zzzeek/sqlalchemy/pull/264. Just an FYI, and we don't need to wait for that PR to merge this one.

File: ibis/sql/postgres/api.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    Create an Ibis client connected to a PostgreSQL database.
+
+    Multiple database files can be created using the attach() method
+
+    Parameters
+    ----------
+    path : string, default None
+        File path to the PostgreSQL database file. If None, creates an
+        in-memory transient database and you can use attach() to add more files
+    create : boolean, default False
```
Comment (wesm) [inline ibis/sql/postgres/api.py]: cruft from sqlite

File: ibis/sql/postgres/client.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+class PostgreSQLClient(alch.AlchemyClient):
+
+    """
+    The Ibis PostgreSQL client class
+    """
+
+    dialect = PostgreSQLDialect
+    database_class = PostgreSQLDatabase
+
+    def __init__(self, path=None, create=False):
```
Comment (wesm) [inline ibis/sql/postgres/client.py]: can delete the `create` cruft

File: ibis/sql/postgres/client.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        raise NotImplementedError
+
+    def set_database(self):
+        raise NotImplementedError
+
+    @property
+    def client(self):
+        return self
+
+    def table(self, name, database=None):
```
Comment (wesm) [inline ibis/sql/postgres/client.py]: database argument unused

File: ibis/sql/postgres/compiler.py [hunk 5]
```diff
@@ -0,0 +1,519 @@
...
+import ibis.expr.operations as ops
+
+
+import ibis.sql.alchemy as alch
+_operation_registry = alch._operation_registry.copy()
+
+
+# TODO: substr and find are copied from SQLite, we should really have a
+# "base" set of SQL functions that are the most common APIs across the major
+# RDBMS
```
Comment (wesm) [inline ibis/sql/postgres/compiler.py]: definitely agree, for any functions that are identical in more than one place

File: ibis/sql/postgres/compiler.py [hunk 6]
```diff
@@ -0,0 +1,519 @@
...
+    op = expr.op()
+
+    klass = _cumulative_to_reduction[type(op)]
+    new_op = klass(*op.args)
+    new_expr = expr._factory(new_op, name=expr._name)
+
+    if type(new_op) in translator._rewrites:
+        new_expr = translator._rewrites[type(new_op)](new_expr)
+
+    return L.windowize_function(new_expr, win)
```
Comment (cpcloud) [inline ibis/sql/postgres/compiler.py]: This seems spurious. I'm not importing `L` anywhere, so I think this code doesn't get used. This should also go in the refactor of base functions.

File: ibis/sql/postgres/compiler.py [hunk 7]
```diff
@@ -0,0 +1,519 @@
...
+    ops.StandardDev: _variance_reduction('stddev'),
+
+    # now is in the timezone of the server, but we want UTC
+    ops.TimestampNow: lambda *args: sa.func.timezone('UTC', sa.func.now()),
+    ops.WindowOp: _window,
+    ops.CumulativeOp: _window,
+})
+
+
+def add_operation(op, translation_func):
```
Comment (cpcloud) [inline ibis/sql/postgres/compiler.py]: Not sure why this is here, probably something left over from the copypaste.

**Comment by wesm** [comment]
Left minor comments; I'll run the data load and test suite tomorrow to make sure all checks out and then merge. This is great stuff, will want to do a 0.8.0 and maybe get some more postgres users hacking on this! Would be great to see Redshift and Vertica dialects based off a common pg-like base

**Comment by cpcloud** [comment]
Cool, thanks for the review!

**Pull request closed without merging**